File: tools/oracle/l6clear.mjs
// Port side of tools/oracle/l6clear.py: does clearing LEVEL 6 hand over to
// level 7, and on which frame?
//
// This is the softlock. loc_00_35E8's default arm is `LD C,$01 / JP
// loc_00_2820` -- column 1 of the 0:$286D pair, i.e. the TOP exit -- and level
// 6 is the only level that reaches it (right = $FF, top = $07). main.js used
// to guard the handoff on exitRight, found $FF, wrote no next level, and left
// the cleared vehicle stage running forever.
//
// Drives the real enemy driver, the real $C740 countdown, the real level-6
// fanfare and the real clearLevel, then replays the two lines of main.js's
// step() that consume the request (the same shape flowdiff.mjs uses -- the
// frame loop itself needs a canvas).
//
//   node tools/oracle/l6clear.mjs [--kill 20] [--frames 1400]

import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath, pathToFileURL } from 'node:url';
import { imp, installFetchShim } from './_env.mjs';

installFetchShim();

const { createState } = await imp('src/state.js');
const { makeTunables } = await imp('src/tunables.js');
const { initLevel, clearLevel } = await imp('src/level.js');
const { loadManifest, loadPlayerTiles } = await imp('src/assets.js');
const { tick } = await imp('src/main.js');
const { resolveLoadout } = await imp('src/mods.js');
const { effects } = await imp('src/effects.js');
const { GAMEPLAY_PALETTES } = await imp('src/state.js');
const { createFramebuffer, renderFrame } = await imp('src/render/renderer.js');

const argv = process.argv.slice(2);
const arg = (n, d) => { const i = argv.indexOf('--' + n); return i >= 0 ? argv[i + 1] : d; };
const killAt = parseInt(arg('kill', '20'), 10);
const frames = parseInt(arg('frames', '1400'), 10);

const manifest = await loadManifest();
const playerTiles = await loadPlayerTiles();
const state = createState(makeTunables());
state.loadout = resolveLoadout([]);
await initLevel(state, 6);

console.log(`level ${state.level.number}  bossId=${state.level.bossId}`
  + `  exitRight=$${state.level.exitRight.toString(16).toUpperCase()}`
  + `  exitTop=$${state.level.exitTop.toString(16).toUpperCase()}`
  + `  vehicle HP=$${state.enemies[0][0x16].toString(16).toUpperCase()}`);

let clearedAt = null;
let handoff = null;
for (let f = 1; f <= frames; f++) {
  if (f === killAt) state.enemies[0][0x16] = 0;     // 1:$4E82's own trigger
  tick(state, manifest, playerTiles);
  if (state.flow.levelCleared === 1 && clearedAt === null) {
    clearedAt = f;
    // --- main.js step()'s clear arm, verbatim -----------------------------
    state.flow.levelCleared = 2;
    const next = clearLevel(state);
    console.log(`f${String(f).padStart(4)}  levelCleared raised;`
      + ` clearLevel -> ${JSON.stringify(next)}`);
    if (next.to === 'transition' && next.exit !== undefined
        && next.exit !== 0xFF && next.exit !== 0xFE) {
      state.flow.nextLevel = next.exit;
    }
    handoff = state.flow.nextLevel;
    break;
  }
  if (effects(state).countdown !== 0xFF && f % 100 === 0) {
    console.log(`f${String(f).padStart(4)}  $C740=$`
      + effects(state).countdown.toString(16).toUpperCase().padStart(2, '0')
      + `  phase=${effects(state).phase} stage=${effects(state).stage}`);
  }
}

if (clearedAt === null) {
  console.log(`\nFAIL: the port never raised flow.levelCleared in ${frames} frames`);
  process.exit(1);
}
if (!handoff) {
  console.log('\nFAIL: SOFTLOCK -- the clear was raised but no next level was '
    + 'written, so the empty level 6 keeps running');
  process.exit(1);
}
// The frame loop then loads it. BOTH lines, in main.js's order -- the palette
// restore is not decoration: level 6's fanfare is a 33-frame fade to WHITE
// ($FFAD/$FFAE/$FFAF walk E4 -> 90 -> 40 -> 00) and this is the only clear that
// leaves through a transition, so without it level 7 plays under a blank
// screen. MEASURED on the cartridge: $FFB0 becomes 07 at f181 and E4/E4/C4 are
// rewritten on that same frame.
Object.assign(state.video, GAMEPLAY_PALETTES);
await initLevel(state, handoff, { transition: true });
console.log(`\nflow.nextLevel = ${handoff}; after the transition load,`
  + ` level = ${state.level.number}`);

// AND THE SCREEN HAS TO SHOW SOMETHING.
//
// This check exists because three separate harnesses drove this exact handover
// and every one reported PASS while the game rendered a solid white frame. They
// asked whether frames RENDERED, never whether they rendered a PICTURE -- which
// is docs/03's "byte-exact data is not a correct picture" in a new costume. The
// reported symptom was "the boss explodes, the screen fades to white, and then
// we softlock": nothing froze and nothing threw, the game was just invisible.
const fb = createFramebuffer();
for (let i = 0; i < 30; i++) {
  state.input.held = 0;
  state.input.pressed = 0;
  tick(state, manifest, playerTiles);
  renderFrame(state, fb);
}
const shades = new Set(fb.shades);
console.log(`palettes: bgp=$${state.video.bgp.toString(16).toUpperCase()}`
  + ` obp0=$${state.video.obp0.toString(16).toUpperCase()}`
  + ` obp1=$${state.video.obp1.toString(16).toUpperCase()}`
  + `   framebuffer shades: ${[...shades].sort().join(',')}`);

const ok = handoff === 7 && state.level.number === 7 && shades.size > 1;
if (shades.size <= 1) {
  console.log('\nFAIL: SOFTLOCK -- level 7 loaded and is running, but every pixel'
    + ' is one shade.\n      The fade-out palette was never restored, so the game'
    + ' is invisible.\n      main.js step(): the flow.nextLevel arm must'
    + ' Object.assign(state.video, GAMEPLAY_PALETTES).');
}
console.log(ok
  ? 'PASS - level 6 hands over to level 7 AND the screen is visible'
  : 'FAIL');
process.exit(ok ? 0 : 1);

File: tools/oracle/l6clear.py
#!/usr/bin/env python3
"""Where does clearing LEVEL 6 send the cartridge?

loc_00_35E8 dispatches levels 4, 8, $0B and $0E and falls through for
everything else into $35FA: `XOR A -> $C73E`, `$C740 = $FF`, `LD C,$01`,
`JP loc_00_2820`.  C is the COLUMN of the 0:$286D exit pair, and 1 is the TOP
exit -- not the right-hand one $1745 loads when the player walks off an edge.

Level 6 is the only level that can reach that arm: it needs a non-zero $C73E
(only levels 4, 6, 8, $0B and $0E have one) and the other four are dispatched
above it.  Its 0:$286D row is right = $FF, top = $07.  So the port's guard on
exitRight found $FF, never wrote a next level, and the cleared vehicle stage
ran forever -- the game could not be finished past level 6 by any route.

Kills the vehicle the way the last hit does, by zeroing its own HP byte
(enemy record 0 at $C268 + $16), and then just watches $FFB0.

  python tools/oracle/l6clear.py
"""
import os
import sys

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from pyboy import PyBoy                                     # noqa: E402
from trace import ROM, FRAME_END, boot_to_gameplay          # noqa: E402

SITES = [
    (0x34D0, 'loc_00_34D0  victory fanfare'),
    (0x34E7, 'loc_00_34E7  level 6 skips the fanfare'),
    (0x35E8, 'loc_00_35E8  the clear dispatch'),
    (0x35FA, '  $35FA      the DEFAULT arm (C = 1, top exit)'),
    (0x360F, '  $360F      route bit 0'),
    (0x2820, 'loc_00_2820  the exit table'),
    (0x2834, '  $2834      $FFB0 <- the chosen exit'),
    (0x285B, '  $285B      $FE: fall back in from the top'),
    # NOT $04BB -- trace.py's boot_to_gameplay already owns a hook there and
    # PyBoy allows only one per (bank, address).
]


def main():
    pyboy = PyBoy(ROM, window='null', sound_emulated=False)
    pyboy.set_emulation_speed(0)
    m = pyboy.memory

    fc = {'n': 0}
    log = []
    for addr, label in SITES:
        pyboy.hook_register(
            0, addr,
            lambda _, l=label: log.append((fc['n'], l, m[0xFFB0], m[0xC740])),
            None)
    pyboy.hook_register(0, FRAME_END, lambda _: fc.__setitem__('n', fc['n'] + 1),
                        None)

    boot_to_gameplay(pyboy, level=6)
    for n in ('start', 'a', 'b', 'left', 'right', 'up', 'down'):
        pyboy.button_release(n)
    if m[0xFFB0] != 6:
        raise SystemExit(f'FAIL: wanted level 6, got ${m[0xFFB0]:02X}')

    base = fc['n']
    log.clear()
    print(f'level ${m[0xFFB0]:02X}  $C73E={m[0xC73E]:02X}  '
          f'vehicle HP=${m[0xC268 + 0x16]:02X}')

    killed = False
    arrived = None
    for _ in range(1400):
        f = fc['n'] - base
        if not killed and f >= 20:
            killed = True
            m[0xC268 + 0x16] = 0            # 1:$4E82's own trigger
        pyboy.tick(1, False)
        if m[0xFFB0] != 6 and arrived is None:
            arrived = fc['n'] - base
            for _ in range(60):
                pyboy.tick(1, False)
            break

    for f, label, lvl, c740 in log:
        print(f'  f{f - base:4d}  {label:44s} $FFB0=${lvl:02X} $C740=${c740:02X}')
    if arrived is None:
        raise SystemExit('FAIL: the cartridge never left level 6')
    print(f'\n$FFB0 left 6 at frame {arrived}; settled at ${m[0xFFB0]:02X}')
    print(f'0:$286D row for level 6: right=${m[0x286D + 10]:02X} '
          f'top=${m[0x286D + 11]:02X}')
    pyboy.stop(save=False)


if __name__ == '__main__':
    main()
